### PR TITLE
Direct slater sampler

### PR DESCRIPTION
**Pull request by jrm874**

### CODE REVIEW

File: python/ffsim/states/slater.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            generator or ``None``.
+    Returns:
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    position = []
+    marginal = []
+    position.append(
+        rng.choice(np.arange(norb), size=1, p=np.real(np.diag(rdm) / nelec))[0]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:134]:
```suggestion
        rng.choice(norb, p=np.real(np.diag(rdm)) / nelec)
```

File: python/ffsim/states/slater.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    )
+    marginal.append(np.real(np.diag(rdm) / nelec)[position[0]])
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:136]: For `position` and `marginal`, instead of initializing them to `[]` and appending a single item, just initialize them with the item, e.g. `[item]`.

File: python/ffsim/states/slater.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        u_conditionals, marginals = _generate_conditionals_unormalized(
+            rdm, pos_array, empty_orbitals, marginal[-1]
+        )
+
+        conditionals = np.real(u_conditionals)
+        conditionals /= np.sum(conditionals)
+
+        n_empty_orbitals = len(empty_orbitals)
+
+        index = rng.choice(np.arange(n_empty_orbitals), size=1, p=conditionals)[0]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:152]:
```suggestion
        index = rng.choice(len(empty_orbitals), p=conditionals)
```

File: python/ffsim/states/slater.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        new_pos = empty_orbitals[index]
+
+        position.append(new_pos)
+        position = list(np.sort(np.array(position)))
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:156]:
```suggestion
        position.sort()
```

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: tests/python/states/slater_test.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    """Test Slater determinant sampler."""
+
+    rng = np.random.default_rng(1234)
+    n_samples = 3000
+    n_a, n_b = nelec
+
+    rotation_a = ffsim.random.random_unitary(norb, seed=rng)
+    rotation_b = ffsim.random.random_unitary(norb, seed=rng)
+    rdm_a = ffsim.slater_determinant_rdms(norb, range(n_a), rotation_a, rank=1)
+    rdm_b = ffsim.slater_determinant_rdms(norb, range(n_b), rotation_b, rank=1)
```
Comment (kevinsung) [inline tests/python/states/slater_test.py:43]:
```suggestion
    rdm_a, rdm_b = ffsim.slater_determinant_rdms(norb, (range(n_a), range(n_b)), (rotation_a, rotation_b))
```

File: tests/python/states/slater_test.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    for i, bts in enumerate(bts_matrix):
+        string = np.array2string(bts, separator="")[1:-1]
+        index = ffsim.strings_to_indices([string], norb, nelec)[0]
+        indices[i] = index
+
+    unique_indices, counts = np.unique(indices, return_counts=True)
+    if isinstance(nelec, tuple):
+        probabilities = np.zeros(math.comb(norb, nelec[0]) * math.comb(norb, nelec[1]))
+    else:
+        probabilities = np.zeros(math.comb(norb, nelec))
```
Comment (kevinsung) [inline tests/python/states/slater_test.py:24]:
```suggestion
    probabilities = np.zeros(ffsim.dim(norb, nelec))
```
This works for both the spinful and spinless cases.

Comment (kevinsung) [inline tests/python/states/slater_test.py:24]:
```suggestion
    probabilities = np.zeros(ffsim.dim(norb, nelec))
```
This works for both the spinful and spinless cases.

File: tests/python/states/slater_test.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+@pytest.mark.parametrize("norb, nelec", ffsim.testing.generate_norb_nocc(range(1, 5)))
+def test_slater_sampler_spinless(norb: int, nelec: int):
+    """Test Slater determinant sampler (spinless case)."""
+
+    rng = np.random.default_rng(1234)
+    n_samples = 3000
+    rotation = ffsim.random.random_unitary(norb, seed=rng)
+    rdm = ffsim.slater_determinant_rdms(norb, range(nelec), rotation, rank=1)
+    test_distribution = (
+        np.absolute(ffsim.slater_determinant(norb, (range(nelec), []), rotation)) ** 2
```
Comment (kevinsung) [inline tests/python/states/slater_test.py:66]:
```suggestion
        np.absolute(ffsim.slater_determinant(norb, range(nelec), rotation)) ** 2
```

File: python/ffsim/states/slater.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+import numpy as np
+
+
+def sample_slater(
+    rdm: np.ndarray | tuple[np.ndarray, np.ndarray],
+    shots: int,
+    seed: np.random.Generator | int | None = None,
+) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply python/ffsim/states/slater.py:10]: Javier, don't worry about this for now. I plan to add helper functions to convert between the different bitstring types.

[the remark this answers is not in the document]
Comment (kevinsung) [reply python/ffsim/states/slater.py:10]: Ok, I've added the helper functions. See [sample_state_vector](https://github.com/qiskit-community/ffsim/blob/ebf393d85a91c4d7dbed8c97a2a8de8a8227bf2f/python/ffsim/states/states.py#L414) for an example of usage. Please make the function signature of `sample_slater` match.

File: python/ffsim/states/slater.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+            generator or ``None``.
+
+    Returns:
+        A 2D Numpy array with samples of electronic configurations.
+        Each row is a sample.
+    """
+
+    rng = np.random.default_rng(seed)
+    norb, _ = rdm.shape
+    nelec = round(np.real(np.sum(np.diag(rdm))))
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:290]: Instead of computing `nelec`, have this function take it as input. Then, we can handle the `np.zeros` and `np.ones` cases here, based on the value of `nelec`. This will make the code in the main function much cleaner.

File: python/ffsim/states/slater.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+            in the positions determined by ``pos_array``.
+    Returns:
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:182]: Unnecessary whitespace.

File: python/ffsim/states/slater.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    conditionals = np.zeros(len(empty_orbitals), dtype=complex)
+    marginals = np.zeros(len(empty_orbitals), dtype=complex)
+
+    for i, orbital in enumerate(empty_orbitals):
+        new_pos_array = np.append(pos_array, [orbital])
+        rest_rdm = rdm[new_pos_array, :]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:188]: Unnecessary colon.

File: python/ffsim/states/slater.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        nelec: Number of electrons.
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+    Returns:
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    position = [rng.choice(norb, p=np.real(np.diag(rdm)) / nelec)]
+    marginal = [np.real(np.diag(rdm) / nelec)[position[0]]]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:217]:
```suggestion
    probs = np.diag(rdm).real / nelec
    position = [rng.choice(norb, p=probs)]
    marginal = [probs[position[0]]]
```

File: python/ffsim/states/slater.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, pos_array: np.ndarray, empty_orbitals: np.ndarray, marginal: float
+) -> tuple[np.ndarray, np.ndarray]:
+    """Generates the conditional and marginal probabilities for adding a particle
+    to the available empty orbitals.
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:162]: Docstring summary should fit on one line. Extra information can go in the main body.

File: python/ffsim/states/slater.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+            must be equal to ``numpy.arange(num_orbitals)``.
+        marginal: The marginal probability associated to having the particles
+            in the positions determined by ``pos_array``.
+    Returns:
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:180]: Unnecessary whitespace. Did the formatter not complain?

File: python/ffsim/states/slater.py [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+        ), convert_bitstring_type(
+            sampled_configuration_a[:, orbs_a],
+            BitstringType.BIT_ARRAY,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, pos_array: np.ndarray, empty_orbitals: np.ndarray, marginal: float
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:159]: Using Numpy arrays for `pos_array` and `empty_orbitals` seems like overkill. It would be simpler and likely more performant to just use Python lists.

File: python/ffsim/states/slater.py [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+    """
+    rng = np.random.default_rng(seed)
+
+    position = [rng.choice(norb, p=np.real(np.diag(rdm)) / nelec)]
+    marginal = [np.real(np.diag(rdm) / nelec)[position[0]]]
+
+    for k in range(nelec - 1):
+        pos_array = np.array(position, dtype=int)
+
+        empty_orbitals = np.setdiff1d(np.arange(norb), pos_array, assume_unique=True)
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:222]: As I mention above, I think we should simply use Python lists for `pos_array` and `empty_orbitals` instead of Numpy arrays.

File: python/ffsim/states/slater.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+
+    rng = np.random.default_rng(seed)
+    norb, _ = rdm.shape
+    nelec = round(np.real(np.sum(np.diag(rdm))))
+    positions = np.zeros((shots, nelec), dtype=int)
+
+    for i in range(shots):
+        positions[i] = _autoregressive_slater(rdm, norb, nelec, rng)
+
+    return _positions_to_bit_array(positions, norb)
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:296]: Let's just return integers, and leave the rest to the conversion functions. I think this should work
```python
return [sum(1 << orb for orb in orbs) for orbs in positions]
```
Also, `samples` might be a more appropriate variable name than `positions`. I'm thinking of each sample as a list of orbitals, i.e., positions.

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: python/ffsim/states/slater.py [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+        ), convert_bitstring_type(
+            strings_b,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, sample: list[int], empty_orbitals: set[int], marginal: float
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:158]: You have `empty_orbitals: set[int]` here but it looks like you actually pass it as a list below.

File: python/ffsim/states/slater.py [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    probs = np.diag(rdm).real / nelec
+    sample = [rng.choice(norb, p=probs)]
+    marginal = [probs[sample[0]]]
+
+    for k in range(nelec - 1):
+        empty_orbitals = list(set(range(norb)).difference(set(sample)))
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:219]:
```suggestion
    all_orbs = set(range(norb))
    for k in range(nelec - 1):
        empty_orbitals = list(all_orbs.difference(sample))
```
Create `set(range(norb))` outside the loop to avoid recreating it. Also, the `difference` method accepts any iterable (it doesn't have to be a set).

File: python/ffsim/states/slater.py [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+        )
+
+        conditionals = np.real(u_conditionals)
+        conditionals /= np.sum(conditionals)
+
+        index = rng.choice(len(empty_orbitals), p=conditionals)
+
+        new_sample = empty_orbitals[index]
+
+        sample.append(new_sample)
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:231]:
```suggestion
        sample.append(empty_orbitals[index])
```

File: python/ffsim/states/slater.py [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+        sample.sort()
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:232]: Is sorting necessary?

File: python/ffsim/states/slater.py [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+        A 2D Numpy array with samples of electronic configurations.
+        Each row is a sample.
+    """
+
+    rng = np.random.default_rng(seed)
+    norb, _ = rdm.shape
+    samples = []
+
+    for i in range(shots):
+        samples.append(_autoregressive_slater(rdm, norb, nelec, rng))
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:269]:
```suggestion
    samples = [_autoregressive_slater(rdm, norb, nelec, rng) for _ in range(shots)]
```

File: tests/python/states/slater_test.py [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    """Test Slater determinant sampler."""
+
+    rng = np.random.default_rng(1234)
+    shots = 3000
+    n_a, n_b = nelec
+
+    rotation_a = ffsim.random.random_unitary(norb, seed=rng)
+    rotation_b = ffsim.random.random_unitary(norb, seed=rng)
+    rdm_a, rdm_b = ffsim.slater_determinant_rdms(
+        norb, (range(n_a), range(n_b)), (rotation_a, rotation_b)
```
Comment (kevinsung) [inline tests/python/states/slater_test.py:35]: Instead of only testing the occupied orbitals being `(range(n_a), range(n_b))`, let's iterate over all the possibilities. This is more thorough and will cover edge cases. For an example of this, see:
https://github.com/qiskit-community/ffsim/blob/57bb49f7cd4a52b5c36cacea9d96b7c37d25c034/tests/python/gates/diag_coulomb_test.py#L159-L162

Same comment applies for the spinless test.

File: tests/python/states/slater_test.py [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    """Test Slater determinant sampler (spinless case)."""
+
+    rng = np.random.default_rng(1234)
+    shots = 3000
+    rotation = ffsim.random.random_unitary(norb, seed=rng)
+    rdm = ffsim.slater_determinant_rdms(norb, range(nelec), rotation, rank=1)
+    test_distribution = (
+        np.absolute(ffsim.slater_determinant(norb, range(nelec), rotation)) ** 2
+    )
+    print(test_distribution)
```
Comment (kevinsung) [inline tests/python/states/slater_test.py:81]: No print statements please.

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: python/ffsim/states/slater.py [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+        n_a, n_b = nelec
+        norb, _ = rdm_a.shape
+
+        if orbs is None:
+            orbs = (range(norb), range(norb))
+        orbs_a = orbs[0]
+        orbs_b = orbs[1]
+
+        if n_a == 0:
+            strings_a = [0 for i in range(shots)]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:110]:
```suggestion
            strings_a = [0] * shots
```

File: python/ffsim/states/slater.py [hunk 26]
```diff
@@ -0,0 +1,271 @@
...
+        elif n_a == norb:
+            strings_a = [sum(1 << j for j in range(norb)) for i in range(shots)]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:112]:
```suggestion
            strings_a = [(1 << norb) - 1] * shots
```

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: python/ffsim/states/slater.py [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+    conditionals = np.zeros(len(empty_orbitals), dtype=complex)
+    marginals = np.zeros(len(empty_orbitals), dtype=complex)
+
+    for i, orbital in enumerate(empty_orbitals):
+        new_sample = copy.deepcopy(sample)
+        new_sample.append(orbital)
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:185]:
```python
new_sample = sample + [orbital]
```
is simpler

File: tests/python/states/slater_test.py (hunk 23 above)
Comment (jrm874) [reply tests/python/states/slater_test.py:35]: Just added it! The tests are a bit slow at the moment

File: python/ffsim/states/slater.py [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+        rest_rdm = rdm[new_sample, :]
+        rest_rdm = rest_rdm[:, new_sample]
+        marginals[i] = np.linalg.det(rest_rdm)
```
[the remark this answers is not in the document]
Comment (jrm874) [reply python/ffsim/states/slater.py:188]: I just did that. It complains that we are casting a complex number into a real one. This happens because the RDM is allowed ot be complex, however since it is a Hermitian matrix, its determinant is guaranteed to be real.

File: tests/python/states/slater_test.py (hunk 23 above)
Comment (kevinsung) [reply tests/python/states/slater_test.py]: How long do they take to run?

File: python/ffsim/states/slater.py [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+            generator or ``None``.
+
+    Returns:
+        A 2D Numpy array with samples of electronic configurations.
+        Each row is a sample.
+    """
+    norb, _ = rdm.shape
+
+    if nelec == 0:
+        return [0 for i in range(shots)]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:238]:
```suggestion
        return [0] * shots
```

File: python/ffsim/states/slater.py [hunk 30]
```diff
@@ -0,0 +1,244 @@
...
+    if nelec == norb:
+        return [sum(1 << j for j in range(norb)) for i in range(shots)]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:240]:
```suggestion
        return [(1 << norb) - 1] * shots
```

File: python/ffsim/states/slater.py [hunk 31]
```diff
@@ -0,0 +1,244 @@
...
+    rng = np.random.default_rng(seed)
+    samples = []
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:242]:
```suggestion
```

File: python/ffsim/states/slater.py [hunk 32]
```diff
@@ -0,0 +1,244 @@
...
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+    marginals = np.zeros(len(empty_orbitals), dtype=float)
+    for i, orbital in enumerate(empty_orbitals):
+        new_sample = sample.copy()
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:167]:
```suggestion
```

File: python/ffsim/states/slater.py [hunk 33]
```diff
@@ -0,0 +1,244 @@
...
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    probs = np.diag(rdm).real / nelec
+    sample = [rng.choice(norb, p=probs)]
+    marginal = [probs[sample[0]]]
+    all_orbs = set(range(norb))
+    for k in range(nelec - 1):
+        empty_orbitals = list(all_orbs.difference(sample))
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:199]: I realize now that this code can be made more efficient. There is actually no need for the `empty_orbitals` variable here. Just update `all_orbs` at the end of the loop by removing the chosen item. But you should actually rename `all_orbs` to `empty_orbitals`, and you need to remove the first item before the start of the loop.

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: python/ffsim/states/slater.py [hunk 34]
```diff
@@ -0,0 +1,244 @@
...
+    Returns:
+        A Numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    probs = np.diag(rdm).real / nelec
+    sample = [rng.choice(norb, p=probs)]
+    marginal = [probs[sample[0]]]
+    all_orbs = set(range(norb))
+    empty_orbitals = list(all_orbs.difference(sample))
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:197]:
```suggestion
    orb = rng.choice(norb, p=probs)
    sample = [orb]
    marginal = [probs[orb]]
    empty_orbitals = set(range(norb))
    empty_orbitals.remove(orb)
```

File: python/ffsim/states/slater.py [hunk 35]
```diff
@@ -0,0 +1,244 @@
...
+    for k in range(nelec - 1):
+        marginals = _generate_marginals(rdm, sample, empty_orbitals, marginal[-1])
+        conditionals = marginals / marginal[-1]
+        conditionals /= np.sum(conditionals)
+        index = rng.choice(len(empty_orbitals), p=conditionals)
+        sample.append(empty_orbitals[index])
+        marginal.append(marginals[index])
+        empty_orbitals.pop(index)
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:205]:
```suggestion
        empty_orbitals.remove(index)
```

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: python/ffsim/states/slater.py [hunk 36]
```diff
@@ -0,0 +1,228 @@
...
+    rng = np.random.default_rng(seed)
+    empty_orbitals = list(range(norb))
+    probs = np.diag(rdm).real / nelec
+    index = rng.choice(len(empty_orbitals), p=probs)
+    sample = [empty_orbitals[index]]
+    marginal = [probs[index]]
+    empty_orbitals.pop(index)
+    for _ in range(nelec - 1):
+        marginals = _generate_marginals(rdm, sample, empty_orbitals)
+        conditionals = marginals / marginal[-1]
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:198]: This line doesn't do anything because the next line normalizes the conditionals.

File: python/ffsim/states/slater.py [hunk 37]
```diff
@@ -0,0 +1,228 @@
...
+    """
+    rng = np.random.default_rng(seed)
+    empty_orbitals = list(range(norb))
+    probs = np.diag(rdm).real / nelec
+    index = rng.choice(len(empty_orbitals), p=probs)
+    sample = [empty_orbitals[index]]
+    marginal = [probs[index]]
+    empty_orbitals.pop(index)
+    for _ in range(nelec - 1):
+        marginals = _generate_marginals(rdm, sample, empty_orbitals)
```
Comment (kevinsung) [inline python/ffsim/states/slater.py:197]: We can avoid the extra logic for the first sample by making `_generate_marginals` return the correct distribution when `sample` is empty.

**Review by kevinsung — APPROVED** [review]
Thank you Javier!
